rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/uri/jsonrpc/impl/Client.java

```diff
@@ -0,0 +1,116 @@
+import org.rascalmpl.vscode.lsp.uri.jsonrpc.VSCodeUriResolverServer;
+import org.rascalmpl.vscode.lsp.uri.jsonrpc.messages.ISourceLocationChanged;
+import org.rascalmpl.vscode.lsp.uri.jsonrpc.messages.WatchRequest;
+import io.usethesource.vallang.ISourceLocation;
+
+public class Client implements VSCodeUriResolverClient {
+
+    private final Map<ISourceLocation, Watchers> watchers = new ConcurrentHashMap<>();
+    private final Map<String, Watchers> watchersById = new ConcurrentHashMap<>();
+
+    @Override
+    public void emitWatch(ISourceLocationChanged event) {
+        var watch = watchersById.get(event.getWatchId());
+        if (watch != null) {
+            watch.publish(event.translate());
+        }
+    }
+
+    @Override
+    public void addWatcher(ISourceLocation loc, Consumer<ISourceLocationWatcher.ISourceLocationChanged> callback, VSCodeUriResolverServer server) throws IOException {
+        var watch = watchers.get(loc);
+        if (watch == null) {
+            // we got a new one
+            watch = new Watchers();
+            watch.addNewWatcher(callback);
+            watchers.put(loc, watch);
+            watchersById.put(watch.id, watch);
+            try {
+                server.watch(new WatchRequest(loc, watch.id)).join();
+                return;
+            } catch (CompletionException ce) {
+                throw new IOException(ce.getCause());
+            }
+        }
+        watch.addNewWatcher(callback);
+    }
+
+    @Override
+    public void removeWatcher(ISourceLocation loc,
+        Consumer<ISourceLocationWatcher.ISourceLocationChanged> callback,
+        VSCodeUriResolverServer server) throws IOException {
+        var watch = watchers.get(loc);
+        if (watch != null) {
+            if (watch.removeWatcher(callback)) {
+                // last watcher done
+                watchers.remove(loc);
+                if (!watch.callbacks.isEmpty()) {
+                    // not empty anymore, add it back and forget clearing it
+                    watchers.put(loc, watch);
+                    return;
+                }
+                watchersById.remove(watch.id);
+                try {
+                    server.unwatch(new WatchRequest(loc, watch.id)).join();
+                } catch (CompletionException ce) {
+                    throw new IOException(ce.getCause());
+                }
+            }
+        }
+    }
+
+
+
+
+    private static final ExecutorService exec = Executors.newCachedThreadPool(new ThreadFactory() {
+        public Thread newThread(Runnable r) {
+            SecurityManager s = System.getSecurityManager();
+            ThreadGroup group = (s != null) ? s.getThreadGroup() : Thread.currentThread().getThreadGroup();
+            Thread t = new Thread(group, r, "FallbackResolver watcher thread-pool");
+            t.setDaemon(true);
+            return t;
+        }
+    });
+
+    private static class Watchers {
```

Comment:
Please add some source code comments to explain the duplicate management of watchers. In principle the watchers are supposed to be registered on the VScode side, right? Is this copy here because we can not send the callbacks over the wire and so we look them up here?

Reply:
Good point, I'll add some comments. And exactly, we cannot keep track of this on the VSCode side, since we use reference equality for the closures. The same closure instance is used later to "unwatch". Initially I designed something around sending over a Identity hashcode, but that was just messy. Also, this way the jsonrpc bridge is not spammed by the same event multiple times. We do the "broadcast" inside java.

Reply:
ok

Reply:
added comment to explain